### ISSUE TITLE
Steps to complete new dev server

### ISSUE THREAD

**Issue by bradallenfisher** [elmsln/elmsln#1454]
@bradallenfisher 
- [x] rsync files to dev server from current live server
- [x] create all vhost config
- [x] get vmhost to do a csr and build out the cosign stuff
- [x] get vmhost to create wildcard dns on hostname so we can test the system is working
- [x] iron out drush aliases so that we have dev and prod set up on the dev server
- [x] create script that pulls an rsync sans .git when we want to shove files around and uses drush to sync dbs.
- [x] 1x update script that rebuilds all sites / settings.php's
- [x] run all the pieces from jenkins 
@btopro
- [x] config.cfg abstraction to be available to ALL levels of drupal at run time
- [x] ability to override things on an environment / deploy level (layer of abstraction away from config directory)
- [x] rewrite drush-create-site to write those globals into sites.php
- [x] write variable globals into settings.php values
- [x] include in version control _elmsln_env_config BUT IGNORE EVERYTHING BELOW IT (or just make it at run time in install)
- [x] drush aliases need to use the same system as above (@ elmsln should always be local)
- [x] environment specific drush aliases (so we can talk to prod from QA for example)
- [x] auto generate as much of registry keychain as is physically possible
- [x] abstract config/scripts/drush-create-site/config.cfg to be assembled via core/config.cfg (or something like this) which looks in config/scripts/drush-create-site/config.cfg, a new location and custom environment location if it exists

## When all this is working and seems like we have a mirror of Live
- create new issue for making Dev Live.

**Comment by btopro** [comment]
@bradallenfisher for future reference
`:%s/'data-courses.elmsln.local/$GLOBALS['elmslncfg']['serviceprefix'] . 'courses.' . $GLOBALS['elmslncfg']['serviceaddress'] . '/g`

`%s/'courses.elmsln.local/'courses.' . $GLOBALS['elmslncfg']['address'] . '/g`

**Comment by btopro** [comment]
what this looks like currently in sites.php
```
         $GLOBALS['elmslncfg']['serviceprefix'] . 'courses.' . $GLOBALS['elmslncfg']['serviceaddress'] . '.neato' => 'courses/services/' . $GLOBALS['elmslncfg']['host'] . '/neato',
         'courses.' . $GLOBALS['elmslncfg']['address'] . '.neato' => 'courses/' . $GLOBALS['elmslncfg']['host'] . '/neato',
         $GLOBALS['elmslncfg']['serviceprefix'] . 'courses.' . $GLOBALS['elmslncfg']['serviceaddress'] . '.test' => 'courses/services/' . $GLOBALS['elmslncfg']['host'] . '/test',
```

**Comment by btopro** [comment]
service address
```
         $base_url= $GLOBALS['elmslncfg']['protocol'] . '://' . $GLOBALS['elmslncfg']['serviceprefix'] . 'courses.' . $GLOBALS['elmslncfg']['serviceaddress'] . '/neato';
```

**Comment by btopro** [comment]
```
         $base_url= $GLOBALS['elmslncfg']['protocol'] . '://courses.' . $GLOBALS['elmslncfg']['address'] . '/neato';
```

**Comment by btopro** [comment]
@bradallenfisher my side of this done aside from the 1-time rebuild of all sites.php / all settings.php's to follow the new methodology reflected in all new site entries

**Comment by bradallenfisher** [comment]
drush alias generation is not happy with remote execution. If I manually create an aliases.drushrc.php file and hardcode in 

```php
$aliases['prod.coursestest.dmd100'] = array(
  'uri' => 'courses.aanda.psu.edu.dmd100',
  'root' => '/var/www/elmsln/core/dslmcode/stacks/courses',
  'remote-host' => 'theactualhostname',
  'remote-user' => 'ouruser',
  'ssh-options' => '-p NNNN',
);

$aliases['local.coursestest.dmd100'] = array(
  'uri' => 'courses.elmsln-dev.vmhost.psu.edu.dmd100',
  'root' => '/var/www/elmsln/core/dslmcode/stacks/courses',
);
```

I can execute code against the live server. But using the current convention I get a less than helpful 
```shell
Drush command terminated abnormally due to an unrecoverable error. 
```

**Comment by bradallenfisher** [comment]
if we can get this drush alias thing figured out today i can start on these scripts.

**Comment by btopro** [comment]
k, i'll be in after proper cafination

**Comment by btopro** [comment]
```
Array
(
    [root] => /var/www/elmsln/core/dslmcode/stacks/blog
    [uri] => blog.PRODURL.eldig
    [ssh-options] => -p NNNN
    [remote-user] => USER
    [remote-host] => CONNECTIONADDRESS
)
```
Give it a shot now

**Comment by bradallenfisher** [comment]
is the sites.php part of this something that is waiting on a new build? 
on live server, the sites.php files are still the same. I am writing a quick and dirty sed script to get this done.

**Comment by btopro** [comment]
the sed thing is a purity thing, i don't think we should put that into the actual platform as an upgrade script formally so much as just run it against our own instance but it is needed to close this out I believe

**Comment by bradallenfisher** [comment]
oh i see... got it
thanks.

**Comment by bradallenfisher** [comment]
so this is a maintenance window type thing then.

**Comment by btopro** [comment]
I mean you're welcome to run it whenever you want so long as the sed works ;) it will have 0 noticable impact on end users

**Comment by bradallenfisher** [comment]
sites.php done.

**Comment by bradallenfisher** [comment]
this is done... however it creates some new issues. I will file them separately.
